### ISSUE TITLE
How to resume training after interruption?

### ISSUE THREAD

**Issue by iamysy** [venturi123/DRLinFluids#15]
Dear Author,

I hope this message finds you well. I am currently using your model for training and have encountered an issue during the process. When my training is interrupted due to server or other non-programmatic issues, I attempt to load the model from the log and resume training. However, I've noticed that after resuming training, the model's reward or training performance seems to remain at the initial state, showing no signs of improvement.

I have tried various approaches, including saving and loading model, buffer, etc., but the problem persists. Do you have any specific advice or solutions regarding this issue? I truly appreciate your patience, guidance, and assistance.

Thank you！

**Comment by venturi123** [comment]
Hi @iamysy ,

Due to the DRLinFluids Package using Tensorforce and Tianshou as the Backend, the model will be saved according to the different choices. I would like to know if you have tried the following model loading and saving code. Here are two references:
https://tensorforce.readthedocs.io/en/latest/agents/agent.html#loading-and-saving
https://pytorch.org/tutorials/beginner/saving_loading_models.html

**Comment by iamysy** [comment]
Dear Author:
Thank you for your reply, I used your side of the code part in the save and load model part for example:
```
parser.add_argument(
        "--resume",
        default=True,
        action="store_true",
        help="restart"
    )
```
```
    def save_best_fn(policy):
        torch.save(
                    {
                        'model': policy.state_dict(),
                        # 'optim': optim.state_dict(),
                    }, os.path.join(log_path, 'best_model.pth')
                   )
```
`    def save_checkpoint_fn(epoch, env_step, gradient_step):
        # see also: https://pytorch.org/tutorials/beginner/saving_loading_models.html
        torch.save(
            {
                'model': policy.state_dict(),
                # 'optim': optim.state_dict(),
            }, os.path.join(log_path, 'checkpoint.pth')
        )
        pickle.dump(
            train_collector.buffer,
            open(os.path.join(log_path, 'train_buffer.pkl'), "wb")
        )`
`result = offpolicy_trainer(
        policy,
        train_collector,
        test_collector,
        args.epoch,
        args.step_per_epoch,
        args.step_per_collect,
        args.test_num,
        args.batch_size,
        update_per_step=args.update_per_step,
        stop_fn=stop_fn,
        save_best_fn=save_best_fn,
        save_checkpoint_fn=save_checkpoint_fn,
        logger=logger,
        resume_from_log=args.resume,
    )`
I'm missing or unsetting something in the above section, even the `buffer` is stored and imported by me, but the reward is still low after I change the `resume` to `True`!

**Comment by iamysy** [comment]
Dear Author:
Thank you for your reply, I used your side of the code part in the save and load model part for example:

```
parser.add_argument(
        "--resume",
        default=True,
        action="store_true",
        help="restart"
    )
```
```
    def save_best_fn(policy):
        torch.save(
                    {
                        'model': policy.state_dict(),
                        # 'optim': optim.state_dict(),
                    }, os.path.join(log_path, 'best_model.pth')
                   )
```
```
    def save_checkpoint_fn(epoch, env_step, gradient_step):
        # see also: https://pytorch.org/tutorials/beginner/saving_loading_models.html
        torch.save(
            {
                'model': policy.state_dict(),
                # 'optim': optim.state_dict(),
            }, os.path.join(log_path, 'checkpoint.pth')
        )
        pickle.dump(
            train_collector.buffer,
            open(os.path.join(log_path, 'train_buffer.pkl'), "wb")
        )
```
```
    if args.resume:
        # load from existing checkpoint
        print(f"Loading agent under {log_path}")
        ckpt_path = os.path.join(log_path, 'checkpoint.pth')
        if os.path.exists(ckpt_path):
            checkpoint = torch.load(ckpt_path, map_location=args.device)
            policy.load_state_dict(checkpoint['model'])
            # policy.optim.load_state_dict(checkpoint['optim'])
            print("Successfully restore policy and optim.")
        else:
            print("Fail to restore policy and optim.")
        buffer_path = os.path.join(log_path, 'train_buffer.pkl')
        if os.path.exists(buffer_path):
            train_collector.buffer = pickle.load(open(buffer_path, "rb"))
            print("Successfully restore buffer.")
        else:
            print("Fail to restore buffer.")
```
```
result = offpolicy_trainer(
        policy,
        train_collector,
        test_collector,
        args.epoch,
        args.step_per_epoch,
        args.step_per_collect,
        args.test_num,
        args.batch_size,
        update_per_step=args.update_per_step,
        stop_fn=stop_fn,
        save_best_fn=save_best_fn,
        save_checkpoint_fn=save_checkpoint_fn,
        logger=logger,
        resume_from_log=args.resume,
    )
```
I'm missing or unsetting something in the above section, even the buffer is stored and imported by me, but the reward is still low after I change the resume to True!

**Comment by venturi123** [comment]
I go through your functions for saving models and don't find obviously bugs. However, it's hard to say if there is any logic error in the whole training loop. I guess you should first check if the best_model or checkpoint is the real one you want to get. For example, you may save your initial model to `init_model.pth` and then the checkpoint as `checkpoint.pth`. Load the two models and see if the following code will output different results.
```
paras = list(net.parameters())
for num, para in enumerate(paras):
    print('number:',num)
    print(para)
    print('_____________________________')
```

If yes, the model is updated, saved, and loaded correctly. Otherwise, you should check the logic error within the three processes.

**Comment by venturi123** [comment]
Hey there! 

I've noticed this issue hasn't had any activity for a while. I'm going to close it for now to help keep things tidy. 
If you have any more questions or run into any problems, feel free to reopen it or start a new one.

Best,
Qiulei
